### ISSUE TITLE
Truncated text and custom font families

### ISSUE THREAD

**Issue by fabiospampinato** [mermaid-js/mermaid#1540]
**Describe the bug**

I can't get Mermaid to not truncate any part of any rendered text. For some context I'm mainly interested in getting Mermaid to work properly inside [Notable](https://github.com/notable/notable), of which you can find a version shipping with Mermaid v8.5.2 [here](https://github.com/notable/notable-experimental/releases), but I can also reproduce some of these issues in the official live editor.

**Issue**

First of all let's take the following graph as an example:

```mermaid
graph LR
    A{Some Long Text}
    A -->|option number one| B[option number one]
    A -->|option number two| C[option number two]
```

This is what it looks like under my Windows VM:

<img width="1440" alt="Screen Shot 2020-07-14 at 00 14 37" src="https://user-images.githubusercontent.com/1812093/87363497-333b8c80-c569-11ea-827c-c69da5950ad4.png">

The padding is completely off, and for some reason inside the app (which you can kind of see there in the background) the issue is more accentuated 🤷‍♂️.

Inside Notable under macOS or Linux this particular graph looks almost ok though (the "o" in "two" is still partially truncated though):

![image](https://user-images.githubusercontent.com/1812093/87363608-71d14700-c569-11ea-9579-e07eebacadca.png)

I think this might be an issue related to the font family, as when comparing the SVG rendered in the live editor and inside the app I've noticed a different font family is being used.

So I've explicitly set the same exact font family I'm using in the rest of the app via the following settings object:

```js
{
  fontFamily: '-apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif'
}
```

And this is the result:

![image](https://user-images.githubusercontent.com/1812093/87363736-b6f57900-c569-11ea-90e0-83244c319f49.png)

For some reason the fonts that end up being used are different, even though the default ones seem to be inferred from the app anyway, perhaps the logic behind that isn't quite right.

Anyway you can see how pretty much everything gets truncated now for some reason.

Another issue is that using the same font family in the live editor changes nothing, making the live editor useless for testing different font families.

Here there's a more comprehensive corpus of diagrams rendered inside Notable looks like, in case you want to test them yourself, most of them are fine with the default settings under my macOS machine, but under my Windows 10 VM the padding used is absurd:

`````markdown
# Mermaid bug

## One

```mermaid
graph LR
    A{Some Long Text}
    A -->|option number one| B[option number one]
    A -->|option number two| C[option number two]
```

## Two 

```mermaid
graph LR
a[Attitude] --> i[Intention]
sn[Subjective norm] --> i
pbc[Perceived behavioral control] --> i
i --> b[Behavior]
pbc --> b
```

## Three

```mermaid
graph LR
  sys{"System II (deliberative) decision"} -->|yes| act[Attitude activation] --> inf[Attitude influences behavior]
  sys -->|no| acc{Attitude accessibility} -->|low| noact[No attitude activation] --> noinf[Attitude does not influence behavior]
  acc -->|high| act
```

## Four

```mermaid
graph TD
  subgraph Coupling
    A-->B
    C-->B
    B-->D
    D-->B
  end
```

## Five

```mermaid
graph LR
Power_Supply --> Transmitter_A
Power_Supply --> Transmitter_B
Transmitter_A --> D
Transmitter_B --> D
```

## Six

```mermaid
graph TB
  start-->step_one
  start-->testing_a_long_string
  step_one-->step_four
  start-->step_four
  step_four-->finish
  testing_a_long_string-->finish
```
`````

Result: 

![image](https://user-images.githubusercontent.com/1812093/87364052-7ba77a00-c56a-11ea-89e6-36d01ea30201.png)

When using the custom font family provided above way more texts get truncated:

![image](https://user-images.githubusercontent.com/1812093/87364200-e658b580-c56a-11ea-8696-88905565e160.png)

Bottom line is Mermaid seems to handle font families in a very brittle manner, it should always render properly with a default configuration and it should still work properly when custom font families are set.

Any idea on how I can get it to not truncate texts?

**Comment by knsv** [comment]
Thanks for reporting this. They method to get the width of the label is to render it and get the bounding box. After that the width if the box is calculated based in that initial width. If the font for some reason differs between the initial rendering and the final svg the text can be cut. I will try to replicate this error and investigate.

This might be related to issue #1485.

**Comment by knsv** [comment]
I can see the issue inside Notable. Very nice note application btw! I will start looking into #1485 and see if that gives any insights into this as well. 

I am available for close collaboration to get this working properly in Notable. If you are interested in that you can join our slack and ping me, knsv.
(https://join.slack.com/t/mermaid-talk/shared_invite/enQtNzc4NDIyNzk4OTAyLWVhYjQxOTI2OTg4YmE1ZmJkY2Y4MTU3ODliYmIwOTY3NDJlYjA0YjIyZTdkMDMyZTUwOGI0NjEzYmEwODcwOTE)

**Comment by pentacular** [comment]
I've come across this problem in 8.7.0.

I'm using the example to have marked generate output with .mermaid class and then using Mermaid.init(undefined, '.mermaid'); to apply that, given import Mermaid from 'mermaid';

Any idea what's missing here?

**Comment by trusktr** [comment]
I wonder why this happens. I would love to use Mermaid, but it is unusable with this issue.

I tried with all my CSS disabled except for Mermaid's, and it still happens.

I really would love to make some node/flow graphs with this!

**Comment by trusktr** [comment]
This is what I see directly on the Mermaid site:

![Screenshot from 2021-01-10 20-23-28](https://user-images.githubusercontent.com/297678/104147071-c2dbbf80-5381-11eb-9615-265d1d132283.png)

I am using Chromium in Manjaro Linux.

**Comment by shatadrusaha** [comment]
I am having similar issue, when I use mermaid in Rmarkdown. When I run the code as a chunck on RStudio, the image is fine. 
![image](https://user-images.githubusercontent.com/18445132/145239582-1974f05c-8f55-4152-8e7d-7b433f3d272d.png)
But, when I render the markdown document into an html file output the  texts are truncated. I have tried chrome, IE and edge. same problem on all of them. 
![image](https://user-images.githubusercontent.com/18445132/145239866-8a00d52c-6eb9-43ac-a91a-6c97aba4de8b.png)

Is this issue same as this one or a different issue?

Can anyone comment and advise please?

**Comment by sliminality** [comment]
This is also happening in Notion, with lines getting cut off entirely:

<img width="1259" alt="image" src="https://user-images.githubusercontent.com/57372/146667679-8403bd96-6c86-4318-95bc-55d6d70cdcf2.png">

The relevant markup:

```mermaid
flowchart LR
	start --> a1((a1))
	a1((a1)) --> 1("lock()\nold := new\nq := q->next") 
	--> a2((a2)) -->
	2[/if q != NULL/] -- TRUE --> a3((a3)) -->
	3("q->data := new\nunlock()\nnew++") --> a4((a4)) -->
	4[/if new != old/] -- FALSE --> a5((a5)) --> t5("unlock()")
	2 -- FALSE --> a4
	4 -- TRUE --> a1
```

Upon inspection, it appears that the `<foreignObject>` elements are sized significantly smaller than the text `<div>`s they contain.

<img width="550" alt="Screen Shot 2021-12-18 at 11 45 13 PM" src="https://user-images.githubusercontent.com/57372/146667656-65ae0250-3db2-45a2-80e4-b9c74a40158a.png">

**Comment by ggrossetie** [comment]
Maybe a simpler solution would be to be able to configure a padding to account for font rendering discrepancies (between a server and a client)?

**Comment by jgreywolf** [comment]
I am not able to repro this in the live editor any more, nor in my local site. Is the base issue reported still an issue?

**Comment by ggrossetie** [comment]
I believe that Mermaid live is using client side rendering? Yes it's still an issue and an easier fix would probably to allow configuring a padding to account for font rendering discrepancies.
So even if the font is slightly wider and/or taller it will still fit.

**Comment by LukaDarsalia** [comment]
I was able to overcome this issue in Notion by using `<br/>` like this and you can change size of the node by adding dots at the end:
```code
graph TD
   f((<br/>f<br/><br/>....))
   x((<br/>x<br/><br/>....))
   y((<br/>y<br/><br/>....))
	 t2((<br/>t<br/><br/>....))
	 t1((<br/>t<br/><br/>....))

f --> x
f --> y
y --> t1
x --> t2
```
Before: 
<img width="208" alt="Screenshot 2023-05-26 at 03 18 38" src="https://github.com/mermaid-js/mermaid/assets/29729836/6da033a6-ea50-491e-9440-dd554d4167d5">
After:
<img width="214" alt="Screenshot 2023-05-26 at 03 17 16" src="https://github.com/mermaid-js/mermaid/assets/29729836/c1348ed7-bee1-412f-ba69-46e47c3a2e11">

**Comment by SkipperJK** [comment]
> yea, still an issue, especially in notion. The font line breaks for labels especially, but looks fine in mermaid live
> 
> <img alt="Screenshot 2023-05-15 at 1 14 27 PM" width="427" src="https://user-images.githubusercontent.com/966644/238421862-5d39be2d-bfc9-4999-a9f9-f9ccc545a834.png"> <img alt="Screenshot 2023-05-15 at 1 14 21 PM" width="783" src="https://user-images.githubusercontent.com/966644/238421867-319951db-8ba9-4dc4-ba16-b7f92523a704.png">

Use backticks to enclose text. like A["`xxx`"]

**Comment by d33bs** [comment]
Just wanted to mention that I witnessed this challenge as well (outside of Notion) and was able to move forward with non-truncated text rendering using @LukaDarsalia's workaround (adding `<br/><br/>` to the node, for ex: `nodename["a clever description"]` to `nodename["a clever description<br/><br/>"]`)(many thanks for the comment about this!). In my case I didn't need the additional `...` after the breaks.

**Comment by sidharthv96** [comment]
All the given examples are rendering properly now. 

```mermaid
graph TD
f((f)) --> x((x))
f --> y((y))
y --> t1((t1))
x --> t2((t2))
```

# Mermaid bug

## One

```mermaid
graph LR
    A{Some Long Text}
    A -->|option number one| B[option number one]
    A -->|option number two| C[option number two]
```

## Two 

```mermaid
graph LR
a[Attitude] --> i[Intention]
sn[Subjective norm] --> i
pbc[Perceived behavioral control] --> i
i --> b[Behavior]
pbc --> b
```

## Three

```mermaid
graph LR
  sys{"System II (deliberative) decision"} -->|yes| act[Attitude activation] --> inf[Attitude influences behavior]
  sys -->|no| acc{Attitude accessibility} -->|low| noact[No attitude activation] --> noinf[Attitude does not influence behavior]
  acc -->|high| act
```

## Four

```mermaid
graph TD
  subgraph Coupling
    A-->B
    C-->B
    B-->D
    D-->B
  end
```

## Five

```mermaid
graph LR
Power_Supply --> Transmitter_A
Power_Supply --> Transmitter_B
Transmitter_A --> D
Transmitter_B --> D
```

## Six

```mermaid
graph TB
  start-->step_one
  start-->testing_a_long_string
  step_one-->step_four
  start-->step_four
  step_four-->finish
  testing_a_long_string-->finish
```

**Comment by s-h-a-d-o-w** [comment]
@sidharthv96 

It looks like this was either not fully resolved or there has been a regression, see this example: https://codepen.io/the-real_s-h-a-d-o-w/pen/RNbejQq

It's running fine in CodePen but if you download it and open it locally, you will very likely see clipping.

This can be resolved easily by inserting the following in the render function:
```ts
  await new Promise(resolve => {
    document.fonts.ready.then(() => {
      resolve()
    });
  })
```

I thought about just submitting a PR given the simplicity. But there should obviously also be a regression test created, which will take a bit of time. But I'm happy to contribute if you want me to.

**Comment by Sytten** [comment]
I can confirm that there is truncation happening on multi-lines too

**Comment by Capevace** [comment]
<img width="736" height="292" alt="Image" src="https://github.com/user-attachments/assets/d22c5954-a175-4579-83d4-6a7844917ea6" />

Yep, I'm seeing similar cutoffs when using some custom fonts (this is Jetbrains Mono).
